zmap/zgrab2
# add MySQL module

# What it does:
* Add mysql zgrab2 module
* Generic TLS grab
  * Add tls.go, abstracting TLS configuration / scanning / output
  * Let ZGrab drive the TLS handshake
     1. The module's Scan() method uses the mysql driver
     2. If the server supports TLS, use the zgrab2 TLS APIs to do the scan based on the input flags 
* Add zschema schemas
* Add travis integration
  * Pull down mysql docker containers to test against
  * Run zschema validation on output

# How to test
When changes are published, Travis will automatically run the integration tests, but it's straightforward to run them manually:
1. Pull down any requirements (see `before_install` and `before_scripts` in .travis.yml)
  * e.g. https://github.com/jmespath/jp
2. Run `start_mysql.sh`
3. Run `integration_tests.sh`
4. Run `cleanup_mysql.sh`

Or, to directly run zgrab, set up a mysql container (see e.g. `launch_mysql_container.sh`), and run e.g.
* `echo 127.0.0.1 | zgrab2 mysql --heartbleed --keep-client-logs --heartbeat-enabled`

# Notes / Caveats
* The debug-only tags are present in the mysql structs, but not in zcrypto or ssh
  * There is no custom JSON Marshaller to interpret these, either
* Only the mysql module has integration tests
* The scan statuses are currently very rudimentary (just success/fail)
* The TLS command-line flags are almost certainly not in their final form (e.g. array values take a comma-separated list, which is processed via a naive split)
* Future TODO: Include input configuration
  * It seems like this would be a good candidate for a zgrab-level result...i.e. drop the `ScanFlags` next to the `result`?
